### ISSUE TITLE
.bind support

### ISSUE THREAD

**Issue by marcbachmann** [marcbachmann/colt#2]
I'd like to easily bind all the values onto a specific context.
This is an advanced use case and maybe even an anti pattern.
@peyerluk do you think this goes too far? :smile: 

E.g.

``` coffee
describe 'Image API', ->

  before (done) ->
    colt()
    .createUser('user')
    .createRequestForUser('authRequest', 'user')
    .createDocumentForUser('document', {title: 'Foo'}, 'user')
    .bind(this)
    .fire(done)
```

We can also bind specific values:

``` coffee
describe 'Image API', ->

  before (done) ->
    colt()
    .createRequestForUser('authRequest', 'user')
    .createDocumentForUser('document', {title: 'Foo'}, 'user')
    .bind(this, 'user', 'authRequest')
    .fire(done)
```

**Comment by peyerluk** [comment]
I wouldn't use the name `bind` as it may be confusing. I would suggest `.bindResults(this)` instead.

And here is another suggestion to use a `.results` method, that only gets fired in a success case. I'm not sure if it is a good idea though, I start to like `bindResults` better :)

``` coffee
describe 'Image API', ->

  before (done) ->
    colt()
    .createRequestForUser('authRequest', 'user')
    .createDocumentForUser('document', {title: 'Foo'}, 'user')
    .results ({@user, @authRequest}) =>
    .fire(done)
```

Not sure about the name `fire`. I associate it with firing an event. How about `end` like in knex?

**Comment by marcbachmann** [comment]
> I wouldn't use the name bind as it may be confusing.

Yes, I thought the same. But there's no reason to use `.bind`, which doesn't work with this library. And I wanted a short, descriptive name. 

> I would suggest .bindResults(this) instead.

:+1: That's easier to read and easy to search for.

If you only want to get the results, wouldn't be promises a better fit?
We can return one if you call `.catch` or `.then`
however, they don't throw if an error occurs.

> Not sure about the name fire. I associate it with firing an event. How about end like in knex?

I wanted to continue with the naming theme :smile:
There's already a `.end` and `.done` alias

**Comment by peyerluk** [comment]
Regarding:

> If you only want to get the results, wouldn't be promises a better fit?
> and
> There's already a .end and .done alias

I would try to stick with one api. Providing a too flexible api can make it more difficult to learn how it is working by reading examples. Flexibility is nice for power users of a library but will give new users a bumpy learning curve, which I think is not suited for a utility library.

**Comment by marcbachmann** [comment]
Yep, I also like `.end` the most. I'll remove the other things in a future version.
Should we execute the `bindResults` method in the same sequence as it gets registered, or at the end?

So if somebody calls 

``` js
colt().createUser('user').bindResults(this).createDocument('document')
```

`document` doesn't get registered

**Comment by marcbachmann** [comment]
This is a bit implicit, but we can also set the `bindResults` by passing the context to the instance constructor. e.g. `colt(this)`

**Comment by marcbachmann** [comment]
Oh, now I understand what you wanted the `.results(function(results){})`  for.
It doesn't imply that the `.done` doesn't need to be called.
But it's just adding a new function to the sequence, which allows you to add a function to the run sequence.
You probably also want to be able to execute `.results` multiple times. Hmm, do you have a better recommendation for that method name? I'd like to add something like that, but I didn't find a name that wouldn't result in confusion. Maybe we can allow the `.end` to be chained and we can keep the `.fire` or `.done` alias and use them in between invocations.

**Comment by marcbachmann** [comment]
I think `tap` is the word we were searching for.

`colt().tap(function (results)})`
and
`colt().tap('specific', function (specific) {})`

Should `tap` alter a value when somebody returns a value in the invoked function?
I think we should't support that as it would result in strange behavior.

**Comment by marcbachmann** [comment]
`colt().tap(function (results)})`
and
`colt().tap(specific, function (specific) {})`
